include/exadg/incompressible_navier_stokes/spatial_discretization/calculators/wall_shear_stress_calculator.cpp

```diff
@@ -0,0 +1,232 @@
+      "Provide at least write_wall_shear_stress_on_IDs = {dealii::numbers::invalid_boundary_id}.")) bool
+    write_on_all_boundary_IDs = false;
+  if(write_wall_shear_stress_on_IDs.size() == 1)
+  {
+    if(write_wall_shear_stress_on_IDs[0] == dealii::numbers::invalid_boundary_id)
+      write_on_all_boundary_IDs = true;
+  }
+  std::map<dealii::types::boundary_id, bool> write_on_boundary_IDs;
+  for(auto const id : write_wall_shear_stress_on_IDs)
+    write_on_boundary_IDs.insert({id, true});
+
+  unsigned int const dofs_per_component =
+    matrix_free->get_dof_handler(dof_index).get_fe(0).dofs_per_cell / dim;
+  unsigned int const    fe_degree = matrix_free->get_dof_handler(dof_index).get_fe(0).degree;
+  dealii::FESystem<dim> fe_system(dealii::FE_DGQ<dim>(fe_degree), dim);
+
+  dealii::QGaussLobatto<dim - 1> face_quadrature(fe_degree + 1);
+  dealii::QGaussLobatto<dim>     quadrature(fe_degree + 1);
+
+#ifdef DEBUG
+  dealii::FEFaceValues<dim> fe_face_values(*mapping,
+                                           fe_system,
+                                           face_quadrature,
+                                           dealii::update_quadrature_points |
+                                             dealii::update_gradients |
+                                             dealii::update_normal_vectors);
+  dealii::FEValues<dim>     fe_values(*mapping,
+                                  fe_system,
+                                  quadrature,
+                                  dealii::update_quadrature_points);
+#else
+  dealii::FEFaceValues<dim> fe_face_values(
+    *mapping, fe_system, face_quadrature, dealii::update_gradients | dealii::update_normal_vectors);
+#endif
+
+  std::vector<dealii::Tensor<2, dim>> velocity_gradients(fe_face_values.n_quadrature_points);
+  std::vector<dealii::types::global_dof_index> dof_indices(fe_system.dofs_per_cell);
+  const dealii::FEValuesExtractors::Vector     vector(0);
+
+  // fill vector with entries on boundary via cell-loop or prepared container.
+  for(auto const & cell : matrix_free->get_dof_handler(dof_index).active_cell_iterators())
+  {
+    if(cell->is_locally_owned() == false)
+      continue;
+
+    if(cell->at_boundary() == false)
+      continue;
+
+    cell->get_dof_indices(dof_indices);
+
+#ifdef DEBUG
+    fe_values.reinit(cell);
+    double const tol_sqrd = std::pow((cell->diameter() * rel_tol), 2);
+#endif
+
+    for(auto const face : cell->face_indices())
+    {
+      if(cell->at_boundary(face) == false)
+        continue;
+
+      if(write_on_all_boundary_IDs ||
+         write_on_boundary_IDs.find(cell->face(face)->boundary_id()) != write_on_boundary_IDs.end())
+      {
+        fe_face_values.reinit(cell, face);
+        fe_face_values[vector].get_function_gradients(src, velocity_gradients);
+
+        for(auto const i : fe_face_values.quadrature_point_indices())
+        {
+          unsigned int matching_idx = face_to_cell_index[face][i];
+
+#ifdef DEBUG
```

Comment:
do we even want to have this check here? ... could be useful if someone accidentally changes face orientation on the boundary at some point.

Reply:
What do you mean by "someone accidentally changes face orientation on the boundary"? A bug being introduced in deal.II?

Reply:
std::cout output is problematic in ExaDG sections guarded by `debug`, because this could cause failing tests when running in debug mode. However, our policy in ExaDG is that tests should pass in both debug and release mode.

For code review right now, I think it would be easier for me if all debug sections were removed and were realized experimentally on a developer branch for now.

Reply:
If someone is manually changing the face orientation on the outer boundary, this influences the ordering of the points/DoFs -> so we would have non-matching points. @peterum commented on this, that this *should* never be the case, which is why I do the matching during `initialize`.

Regarding the debug sections and std::cout there:
do the checks compare the log files via diff?
If the face orientation is not changed on the boundary (by mistake?!), even debug mode does not output anything - it ONLY CHECKS, if the match was successful, but should still not output anything. So I think if the tests then fail here ... this actually tells us a lot just as we want, right?

Of course one could remove that entirely, but this is kind of a safety measure. We could also always match the points for each face, but that is considerably more effort. And that wall shear stress will lateron be needed for adapting boundary conditions, so it has to be computed at each time step!

So I would actually rather keep it.

Reply:
Ok, I did that now.
So what we do in `initialize` is generate a face index to cell index map and later on, we can use this map to get indices from the face to the DoFs on the cell. 

This assumes that face orientations do not change, and are in fact similar to the first cell encountered on each subdomain. Since I do not know enough regarding changes in orientation or when they might occur (maybe even by accident?), I put the initial node matching into the cell loop to be re-done for each cell all the time in debug mode and Assert if it does not match.

In release mode, I skip this match on each face integration point to the corresponding cell integration point.